### PR TITLE
SQLAlchemy rewrite

### PR DESCRIPTION
**Pull request by RustyBower**

### CODE REVIEW

**Comment by RustyBower** [comment]
I wasn't sure what to do with the requirements file, so please test this out and let me know what I should pin sopel versions to...

This _should_ work with any version of sopel, but please confirm my suspicions :)

File: sopel_modules/github/webhook.py [hunk 1]
```diff
@@ -46,32 +67,16 @@ def setup_webhook(sopel):
     sopel.memory['gh_webhook_server'] = base
     sopel.memory['gh_webhook_thread'] = server
 
-    conn = sopel.db.connect()
-    c = conn.cursor()
 
+    engine = sopel.db.engine
```
Comment (dgw) [inline sopel_modules/github/webhook.py]: Uhhhh, this attribute Should™ be for Sopel's internal use only. The API-defined way of getting at this is [`sopel.db.connect()`](https://github.com/sopel-irc/sopel/blob/514d6b8c70b5853d96dd2093d87eecfd8b176c62/sopel/db.py#L177-L179), which also runs `.connect()` on the engine for you.

File: sopel_modules/github/github.py [hunk 2]
```diff
@@ -426,26 +437,39 @@ def configure_repo_colors(bot, trigger):
...
+               .filter(GithubHooks.repo_name == repo_name) \
+               .one_or_none()
+
+            bot.say("[{}] Example name: {} tag: {} commit: {} branch: {} url: {}".format(
+                    formatting.fmt_repo(repo_name, row),
+                    formatting.fmt_name(trigger.nick, row),
+                    formatting.fmt_tag('tag', row),
+                    formatting.fmt_hash('c0mm17', row),
+                    formatting.fmt_branch('master', row),
+                    formatting.fmt_url('http://git.io/', row)))
```
Comment (dgw) [inline sopel_modules/github/github.py]: Just an itty-bitty thing to tweak As Long As This Line Is Touched Anyway™.

```suggestion
                    formatting.fmt_url('https://git.io/', row)))
```

**Review by dgw — COMMENTED** [review]
tl;dr: `sopel.db.engine` isn't technically part of the API, and shouldn't be used directly by plugins.

**Comment by dgw** [comment]
> This _should_ work with any version of sopel

I doubt it. Sopel before 7.0 (current `master` there) uses SQLite3 directly, not SQLAlchemy, so a lot of this stuff probably won't work properly. Just bump the requirements to `sopel>=7.0,<8` like you did for the YouTube plugin.

File: sopel_modules/github/webhook.py (hunk 1 above)
Comment (RustyBower) [reply sopel_modules/github/webhook.py]: This is what happens when I'm tired and writing code :D

File: sopel_modules/github/github.py (hunk 2 above)
Comment (RustyBower) [reply sopel_modules/github/github.py]: Done.

File: sopel_modules/github/webhook.py (hunk 1 above)
Comment (RustyBower) [reply sopel_modules/github/webhook.py]: Done.

**Comment by RustyBower** [comment]
Refactored all the database connectivity things.

File: sopel_modules/github/webhook.py [hunk 3]
```diff
@@ -102,16 +106,32 @@ def stop(self):
...
-    c.execute('SELECT * FROM gh_hooks WHERE repo_name = ? AND enabled = 1', (repo.lower(), ))
-    return c.fetchall()
+    # Catch any errors connecting to database
+    try:
+        engine = sopel_instance.db.connect()
+    except OperationalError:
+        print("OperationalError: Unable to connect to database.")
+        raise
+
+    session = scoped_session(sessionmaker(bind=engine))
```
Comment (dgw) [inline sopel_modules/github/webhook.py]: Sometimes you use `engine`, sometimes `engine.engine`. Are you sure about this code?

**Review by dgw — CHANGES_REQUESTED** [review]
I see some issues with consistency. Seems to be very confusing naming variables `engine` if they actually hold `Connection` objects?

Also, the repeated try/except for getting a DB session could probably stand to get abstracted out into a function. Maybe a function that takes a `Sopel` (`bot`) object and returns the `session` (or raises an exception if it can't get one)?

I was really hoping to avoid importing SQLAlchemy into this project, but I get the impression that we _have_ to unless the `DB` object API in Sopel itself gets fleshed out a ton.

File: sopel_modules/github/webhook.py (hunk 3 above)
Comment (RustyBower) [reply sopel_modules/github/webhook.py]: Weirdly it works both ways lol

**Comment by RustyBower** [comment]
The colors don't translate into GH comments, but I promise they still look fabulous.
```05:24:24 <@RustyCloud> .gh-hook rustybower/test enable
05:24:24 <sopel> Successfully enabled listening for rustybower/test's events in #bottest.
05:24:24 <sopel> Great! Please allow me to create my webhook by authorizing via this link: https://git.io/Jea5L
05:24:25 <sopel> Once that webhook is successfully created, I'll post a message in here. Give me about a minute or so to set it up after you authorize. You can configure the colors that I use to display webhooks with .gh-hook-color
05:25:30 <sopel> [test] RustyBower: It's not fully shipped until it's fast. (Your webhook is now enabled)
05:27:51 <sopel> [test] RustyBower pushed 1 new commit to master https://git.io/JeVu4
05:27:52 <sopel> test/master 335e2ae Rusty Bower: testing
05:28:00 <@RustyCloud> .gh-hook-color rustybower/test 1 2 3 4 5 6
05:28:00 <sopel> [rustybower/test] Example name: RustyCloud tag: tag commit: c0mm17 branch: master url: https://git.io/
05:28:21 <sopel> [test] RustyBower pushed 1 new commit to master https://git.io/JeVuB
05:28:21 <sopel> test/master 441cae1 Rusty Bower: testing
```

**Comment by dgw** [comment]
Why is the same identical function defined twice? Not to nitpick too much, but that seems very silly. Even sillier than the fact that `setup_webhook()` still has dedicated DB-connection code… which is kind of on me, because I mentioned sessions and that doesn't need one. 😆

> The colors don't translate into GH comments

Indeed, the relative lack of formatting options here vs. on IRC caught me out when trying to write nice PR descriptions for additions to `sopel.tools.formatting` a while back.

File: sopel_modules/github/github.py [hunk 4]
```diff
@@ -118,6 +122,16 @@ def shutdown(sopel):
...
 
 
+def get_db_session(bot):
+    try:
+        engine = bot.db.connect()
+    except OperationalError:
+        print("OperationalError: Unable to connect to database.")
+        raise
+
+    return scoped_session(sessionmaker(bind=engine))
```
Comment (dgw) [inline sopel_modules/github/github.py]: Isn't this pulled in from `.utils` now?

**Review by dgw — COMMENTED** [review]
Not doing a full-on review, but there's one thing I noticed (line-noted). Also, should this not use a similar approach to Sopel's own DB class for thread-safety? That is, with regard to the scoped session management approach you proposed in sopel-irc/sopel#1755.

**Comment by dgw** [comment]
It's no longer _necessary_ to make any changes just to make this plugin keep working—tweaks to Sopel's database API mean this and other plugins that use direct DB access with Sopel 6.x will continue to work with 7.x *if* the default SQLite database backend is in use.

That said, we should still migrate onto SQLAlchemy eventually. It'll give this plugin compatibility with more database types, and as more upstream users take advantage of the new supported backends the restriction to using this plugin with SQLite will become more onerous.

**Comment by dgw** [comment]
I worked on a rebase of this branch, currently stored at `49-rebase` as I'm not 100% confident it's ready to force-push over the PR yet.

Looking into the webhook server getting stalled, first.
